src/lsptoolshost/buildDiagnosticsService.ts

```diff
@@ -0,0 +1,146 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as vscode from 'vscode';
+
+export enum AnalysisSetting {
+    FullSolution = 'fullSolution',
+    OpenFiles = 'openFiles',
+    None = 'none',
+}
+
+export class BuildDiagnosticsService {
+    /** All the build results sent by the DevKit extension. */
+    private _allBuildDiagnostics: Array<[vscode.Uri, vscode.Diagnostic[]]> = [];
+
+    /** The diagnostic results from build displayed by VS Code. When live diagnostics are available for a file, these are errors that only build knows about.
+     * When live diagnostics aren't loaded for a file, then these are all of the diagnostics reported by the build.*/
+    private _diagnosticsReportedByBuild: vscode.DiagnosticCollection;
+
+    constructor(buildDiagnostics: vscode.DiagnosticCollection) {
+        this._diagnosticsReportedByBuild = buildDiagnostics;
+    }
+
+    public clearDiagnostics() {
+        this._diagnosticsReportedByBuild.clear();
+    }
+
+    public async setBuildDiagnostics(
+        buildDiagnostics: Array<[vscode.Uri, vscode.Diagnostic[]]>,
+        buildOnlyIds: string[]
+    ) {
+        this._allBuildDiagnostics = buildDiagnostics;
+        const displayedBuildDiagnostics = new Array<[vscode.Uri, vscode.Diagnostic[]]>();
+        const allDocuments = vscode.workspace.textDocuments;
+
+        this._allBuildDiagnostics.forEach((fileDiagnostics) => {
+            const uri = fileDiagnostics[0];
+            const diagnosticList = fileDiagnostics[1];
+
+            // Check if the document is open
+            const document = allDocuments.find((d) => this.compareUri(d.uri, uri));
+            const isDocumentOpen = document !== undefined ? !document.isClosed : false;
+
+            // Show the build-only diagnostics
+            displayedBuildDiagnostics.push([
+                uri,
+                BuildDiagnosticsService.filerDiagnosticsFromBuild(diagnosticList, buildOnlyIds, isDocumentOpen),
+            ]);
+        });
+
+        this._diagnosticsReportedByBuild.set(displayedBuildDiagnostics);
+    }
+
+    private compareUri(a: vscode.Uri, b: vscode.Uri): boolean {
+        return a.path.localeCompare(b.path, undefined, { sensitivity: 'accent' }) === 0;
```

Comment:
@dibarbet Did you mention something about uri comparision with different cases? Sometimes paths are "c:/users/FileName.cs" and "C:/users/FileName.cs".  It's not safe to do an ignore case compare?

Reply:
Its not safe - some file systems are case sensitive

Reply:
Yes, thank you!